### PR TITLE
Provide EVM builtins `loadimmutable` and `setimmutable`

### PR DESCRIPTION
**Pull request by Marenz**
refs #8551

This is just a draft, tests are missing.
This is just to assure the basic idea of the code changes and whether this can theoretically work like this.

### CODE REVIEW

File: libyul/backends/evm/EVMAssembly.cpp [hunk 1]
```diff
@@ -215,6 +215,16 @@ AbstractAssembly::SubID EVMAssembly::appendData(bytes const&)
...
 }
 
+void EVMAssembly::appendImmutable(std::string const&)
+{
+	yulAssert(false, "LoadImmutable not implemented.");
+}
+
+void EVMAssembly::appendImmutableAssignment(std::string const&)
+{
+	yulAssert(false, "LoadImmutable not implemented.");
```
Comment (chriseth) [inline libyul/backends/evm/EVMAssembly.cpp]:
```suggestion
	yulAssert(false, "SetImmutable not implemented.");
```

File: libyul/backends/evm/EVMDialect.cpp [hunk 2]
```diff
@@ -170,6 +170,45 @@ map<YulString, BuiltinFunctionForEVM> createBuiltins(langutil::EVMVersion _evmVe
 				_assembly.appendInstruction(evmasm::Instruction::CODECOPY);
 			}
 		));
+		builtins.emplace(createFunction(
+			"setimmutable",
+			2,
+			1,
```
Comment (chriseth) [inline libyul/backends/evm/EVMDialect.cpp]:
```suggestion
			0,
```

File: libyul/backends/evm/EVMDialect.cpp [hunk 3]
```diff
@@ -170,6 +170,45 @@ map<YulString, BuiltinFunctionForEVM> createBuiltins(langutil::EVMVersion _evmVe
...
+			SideEffects{},
```
Comment (chriseth) [inline libyul/backends/evm/EVMDialect.cpp]:
```suggestion
			SideEffects{false, false, false, false, true},
```

File: libyul/backends/evm/EVMDialect.cpp [hunk 4]
```diff
@@ -170,6 +170,45 @@ map<YulString, BuiltinFunctionForEVM> createBuiltins(langutil::EVMVersion _evmVe
...
+			{true, false},
+			[](
+				FunctionCall const& _call,
+				AbstractAssembly& _assembly,
+				BuiltinContext&,
+				std::function<void()> _visitArguments
+			) {
+				solAssert(_call.arguments.size() == 2, "");
+
+				_visitArguments();
```
Comment (chriseth) [inline libyul/backends/evm/EVMDialect.cpp]: Hm, we should actually only visit the second argument, right?

File: libyul/backends/evm/EVMDialect.cpp [hunk 5]
```diff
@@ -170,6 +170,45 @@ map<YulString, BuiltinFunctionForEVM> createBuiltins(langutil::EVMVersion _evmVe
...
+				Expression const& arg = _call.arguments.front();
+				YulString identifier = std::get<Literal>(arg).value;
```
Comment (chriseth) [inline libyul/backends/evm/EVMDialect.cpp]: Do you need the local variable?

File: libyul/backends/wasm/WasmDialect.cpp [hunk 6]
```diff
@@ -104,6 +104,8 @@ WasmDialect::WasmDialect()
 
 	addFunction("datasize", {i64}, {i64}, true, {true});
 	addFunction("dataoffset", {i64}, {i64}, true, {true});
+	addFunction("setimmutable", {i32, i32}, {i64}, true, {true, false});
```
Comment (chriseth) [inline libyul/backends/wasm/WasmDialect.cpp]: Maybe better to leave this for now.

File: libyul/backends/evm/EVMDialect.cpp (hunk 4 above)
Comment (Marenz) [reply libyul/backends/evm/EVMDialect.cpp]: Yeah, I thought so too. Maybe I need to modify the parameter to allow specifying which one to visit

**Comment by Marenz** [comment]
Changed code to use a helper function for the `visitArguments` part.

Added the commit of @chriseth 's PR  #8583 

Fixed some cmd line tests that had only whitespace changes.

When running immutable tests with `compileViaYul: also`, we get errors like this:

```
semanticTests/immutable/getter.sol: Exception during test: ../libsolidity/codegen/ir/IRGenerationContext.cpp(86): Throw in function void solidity::frontend::IRGenerationContext::registerImmutableVariable(const solidity::frontend::VariableDeclaration&)
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what: Reserved memory has already been reset.
[solidity::util::tag_comment*] = Reserved memory has already been reset.
```

Will of course squash once approved.

**Comment by chriseth** [comment]
Added this commit to https://github.com/ethereum/solidity/pull/8583

**Pull request closed without merging**